### PR TITLE
Consolidate v4 startup migration notices

### PR DESCRIPTION
**Pull request by logancyang**
Fixes logancyang/obsidian-copilot-preview#287

## Why

A pre-v4 vault can open separate startup dialogs for credentials, commands, system prompts, projects, folder relocation, Miyo, and license handling. Fresh installs can also receive a recommended-command prompt. Those overlapping dialogs obscure which migrations succeeded and what still needs attention.

## What

- Wait for the applicable startup migrations and show their outcomes in one existing `ConfirmModal` with plain text and one **Done** action.
- Keep exact recovery paths and manual steps in that summary, including credentials, unsupported files, folders, Miyo, and a missing legacy license key.
- Remove the fresh-install recommended-command popup; **Generate Default** remains available in Settings.
- Keep network license validation outside the local migration barrier so it cannot delay the summary.

## Non goal

This does not change credential or backup formats, automatically move legacy files, redesign Settings, add migration history, or suppress unrelated runtime errors and genuine license expiry.

## Screenshot

Captured in Obsidian with representative mixed outcomes. The implementation uses the existing confirmation modal rather than adding a new visual component.

Minimal Copilot upgrade summary in Obsidian

## Risk

**High**

| Criterion | Status | Reason |
| --- | :---: | --- |
| No behavior change, or a cosmetic/copy/docs/config change visible where it renders, or deterministic tests cover the changed behavior | ✅ | Coordinator tests cover ordering, rejection fallback, empty startup, formatting, and presentation failure; source tests cover migration and recovery branches |
| A defect would fail CI or be obvious on first use | ✅ | Focused tests run in CI and the one-time summary is immediately visible on upgrade |
| A revert fully restores prior state, including persisted data | ❌ | A device-owned credential-recovery receipt can remain in `data.json`; the credential backup itself remains intact |
| No auth, permissions, secrets, or input-handling surface changes | ❌ | Legacy credential recovery and missing-license handling now report through the combined summary |
| No public API, plugin API, message, or on-disk contract changes | ❌ | Adds optional `_pendingCredentialRecovery` metadata until its instructions are shown |
| No core-path concurrency, async-lifecycle, or state-machine changes | ❌ | Independent layout-ready migrations now settle before one presentation step |
| No hot-path behavior lacks deterministic coverage | ✅ | This is a one-time startup path, and its coordinator and migration branches have deterministic tests |
| No new dependency | ✅ | Dependency manifests are unchanged |
| Human-only behavior stays in one feature area and surfaces quickly | ✅ | Residual visual and ordering risk is confined to the one upgrade summary |

Review: inspect `src/main.ts` — `runStartupMigrations` and `collectLegacyUpgradeRelocation`; `src/services/startupMigration.ts` — `runStartupMigrationSummary` and `formatStartupMigrationSummary`; `src/services/settingsPersistence.ts` — `loadSettingsWithKeychain` and credential recovery; and `src/settings/model.ts` — `_pendingCredentialRecovery` — then run Verification steps 2–5.

## Verification

1. Run `npm run format:check`, `npm run lint`, `npm test -- --runInBand`, `npm run test:mobile-load`, and `npm run review:obsidian`.
2. Open a disposable pre-v4 vault containing legacy credentials, commands, a system prompt, projects, customized Copilot folders, and a prior paid state. Confirm startup shows one summary, with every applicable outcome and recovery path.
3. Force an unsupported command/project and a filesystem failure. Confirm preserved data remains recoverable, and a credential-backup failure leaves `data.json` untouched for retry.
4. Interrupt startup before layout-ready, restart, and confirm credential and folder guidance remains pending until the summary opens. Sync the receipt to a second device and confirm that device neither displays nor clears it.
5. Test a missing legacy license key, an invalid stored key, and stalled online validation. Confirm re-entry guidance does not show an expiry modal, genuine invalid-key handling remains, and network delay never blocks local migration results. Confirm a fresh vault has no startup popup and **Generate Default** remains under Settings → Command.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -486,36 +460,139 @@ export default class CopilotPlugin extends Plugin {
...
+      status: "action-required",
+      summary: "Copilot now keeps its files under one folder. Existing files were not moved.",
+      details: entries.map(
+        ({ label, oldPath, newPath }) => `${label}: move ${oldPath} to ${newPath}.`
+      ),
+    };
+  }
+
+  /** Run all layout-dependent migration work before presenting one summary. */
+  private async runStartupMigrations(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts:488]: Valid — the one-summary invariant and partial-failure settling are only covered at source/result level. Fixing with a focused coordinator test that asserts aggregation, one modal, relocation clearing, and legacy license behavior.

[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts:488]: Fixed in 728e2290 — coordinator coverage now asserts partial-failure settling, retained credentials, relocation clearing, conditional license validation, one complete modal, and the empty current-vault case.

File: src/main.ts [hunk 2]
```diff
@@ -486,36 +460,139 @@ export default class CopilotPlugin extends Plugin {
...
+      }
+    );
+    const relocationTask = settle(this.collectLegacyUpgradeRelocation(), {
+      id: "folders",
+      title: "Copilot folders",
+      status: "error",
+      summary: "Folder destinations could not be prepared. Reload Obsidian to retry.",
+    });
+    const licenseTask = this.isLegacyUpgrade
+      ? checkIsPaidUser().then((isPaid): StartupMigrationItem | null => {
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Valid — a stalled /license request can withhold already-completed one-time recovery results. Fixing by decoupling license validation from the local migration barrier so the summary opens without waiting on network I/O.

[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Fixed in 3cb8d6d7 — online license validation now runs outside the local migration barrier; the migration-specific re-entry action is derived synchronously from a missing legacy key, and the coordinator test keeps validation unresolved while asserting the summary opens.

File: src/services/settingsPersistence.ts [hunk 3]
```diff
@@ -139,6 +143,25 @@ export async function loadSettingsWithKeychain(
...
   keychain.setVaultId(vaultId);
+  let credentialMigration: StartupMigrationItem | null = null;
+
+  const reportCredentialMigration = (): void => {
+    if (credentialMigration && onMigration) {
+      onMigration(credentialMigration);
+    }
+  };
+  const showFallbackNotice = (message: string, duration?: number): void => {
+    if (!onMigration || !credentialMigration) new Notice(message, duration);
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/services/settingsPersistence.ts:186]: Valid — stripping data.json before layout can lose the sole recovery instructions. Fixing with one pending credential-backup receipt in existing settings, cleared only after the summary opens; no migration history or generic report store.

[the remark this answers is not in the document]
Comment (logancyang) [reply src/services/settingsPersistence.ts:186]: Fixed in 4eb040ad — data.json now retains only the backup path/encrypted receipt until the summary opens; the next startup replays it after credentials are stripped, and a modal-open failure leaves it pending.

File: src/main.ts [hunk 4]
```diff
@@ -486,36 +460,143 @@ export default class CopilotPlugin extends Plugin {
...
+    const license: StartupMigrationItem | null = needsLicenseReentry
+      ? {
+          id: "copilot-license",
+          title: "Copilot license",
+          status: "action-required",
+          summary: "Copilot could not restore the previous paid status after the upgrade.",
+          details: ["Re-enter the license key in Copilot Settings to restore paid features."],
+        }
+      : null;
+    if (this.isLegacyUpgrade) void checkIsPaidUser();
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Valid — this background call should preserve the prior invalid-license notification path. Fixing it to pass the plugin App while keeping the network request outside the migration barrier.

File: src/main.ts [hunk 5]
```diff
@@ -486,36 +460,143 @@ export default class CopilotPlugin extends Plugin {
...
-        this.app,
-        () => {},
-        React.createElement(UpgradeRelocationNotice, { entries }),
-        "",
-        "OK",
-        ""
-      ).open();
     }
-
     updateSetting("upgradedToV8FromLegacy", false);
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Valid — the folder guidance can be reconstructed without a new receipt by leaving the existing legacy-upgrade flag set until a non-error folder item is shown. Fixing the clear timing and modal-failure coverage.

[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Fixed in a7b4d5e5 — folder guidance now leaves the existing upgrade flag set until a non-error folder item has been opened in the summary. A modal-open failure keeps both folder and credential recovery pending for restart.

File: src/main.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Fixed in a7b4d5e5 — legacy background validation now calls checkIsPaidUser(this.app), preserving the prior invalid/expired-key notification without adding the network request back to the local migration barrier.

File: src/services/settingsPersistence.ts [hunk 6]
```diff
@@ -160,7 +215,14 @@ export async function loadSettingsWithKeychain(
         // Reason: `enc_*` values cannot be pasted back in, since v4 stores
         // ciphertext verbatim as the key. Those users need fresh keys, so the
         // "copy them across" instruction would send them in circles.
-        new Notice(
+        diskSettings._pendingCredentialRecovery = {
+          path: backup.path,
+          encrypted: backup.encrypted,
+        };
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/services/settingsPersistence.ts:228]: Valid — the backup path is device-local while data.json can sync. Fixing the minimal shape by tagging the single receipt with the existing device-local ID and only reporting/clearing it on its owner device; no per-device report history.

File: src/main.ts [hunk 7]
```diff
@@ -486,36 +460,147 @@ export default class CopilotPlugin extends Plugin {
...
+    const license: StartupMigrationItem | null = needsLicenseReentry
+      ? {
+          id: "copilot-license",
+          title: "Copilot license",
+          status: "action-required",
+          summary: "Copilot could not restore the previous paid status after the upgrade.",
+          details: ["Re-enter the license key in Copilot Settings to restore paid features."],
+        }
+      : null;
+    if (this.isLegacyUpgrade) void checkIsPaidUser(this.app);
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Valid — passing App restored the stored-invalid-key path but also exposed the missing-key branch already represented in the summary. Fixing the background check to run only when a stored key exists.

[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Fixed in 3a5cfe5e — legacy background validation now runs only when a stored key exists, so a missing key produces only the re-entry item in the upgrade summary. Stored invalid keys still validate with the plugin App.

File: src/services/settingsPersistence.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (logancyang) [reply src/services/settingsPersistence.ts:228]: Fixed in 3a5cfe5e — the receipt now carries the existing device-local ID; only its owner reports and clears it, while another device preserves it unchanged. Tests cover local replay, foreign-device preservation, and owner-only clearing.

File: src/main.ts [hunk 8]
```diff
@@ -486,36 +460,153 @@ export default class CopilotPlugin extends Plugin {
...
+      : null;
+    if (this.isLegacyUpgrade && initialSettings.plusLicenseKey) {
+      void checkIsPaidUser(this.app);
+    }
+
+    const [project, commands, prompts, relocation] = await Promise.all([
+      projectTask,
+      commandsTask,
+      promptsTask,
+      relocationTask,
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts]: Won't fix here — issue #287 explicitly excludes a persistent report store/history. These migrations clear legacy settings only after writing recoverable files to deterministic folders, so interruption loses convenience text, not user data; credential and folder receipts remain only where guidance cannot be rederived.

File: src/main.ts [hunk 9]
```diff
@@ -486,36 +462,147 @@ export default class CopilotPlugin extends Plugin {
...
+      status: "action-required",
+      summary: "Copilot now keeps its files under one folder. Existing files were not moved.",
+      details: entries.map(
+        ({ label, oldPath, newPath }) => `${label}: move ${oldPath} to ${newPath}.`
+      ),
+    };
+  }
+
+  /** Run all layout-dependent migration work before presenting one summary. */
+  private async runStartupMigrations(isLegacyUpgrade: boolean): Promise<void> {
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts:492]: Valid — the helper covers sequencing but not the plugin's complete task set and acknowledgement rules. Fixing with required typed migration slots plus a directly tested pure acknowledgement rule; I will not restore the 279-line main-module mock harness.

[the remark this answers is not in the document]
Comment (logancyang) [reply src/main.ts:492]: Fixed in 48ee1cc4 — the coordinator now requires all four migration task slots at compile time, and direct tests cover credential/folder acknowledgement without restoring the 279-line main-module mock harness.
